kaiax/valset/impl/getter_council.go

```diff
@@ -113,33 +113,62 @@ func (v *ValsetModule) getCouncilFromIstanbulSnapshot(targetNum uint64, write bo
 		council, err := v.getCouncilGenesis()
 		return council, 0, err
 	}
+
 	// Load council at the nearest istanbul snapshot. This is the result
 	// applying the votes up to the snapshotNum.
 	snapshotNum := roundDown(targetNum-1, istanbulCheckpointInterval)
-	header := v.Chain.GetHeaderByNumber(snapshotNum)
-	if header == nil {
-		return nil, 0, errNoHeader
-	}
 
-	// Load council at the nearest istanbul snapshot except snapshot num is 0.
-	council := new(valset.AddressSet)
-	if snapshotNum > 0 {
-		council = valset.NewAddressSet(ReadIstanbulSnapshot(v.ChainKv, header.Hash()))
-		if council.Len() == 0 {
-			return nil, 0, ErrNoIstanbulSnapshot(snapshotNum)
+	var council *valset.AddressSet
+	// Try to get from cache first
+	cached, ok := v.councilCache.Get(targetNum - 1)
+	if ok {
+		council = cached.(*valset.AddressSet)
+		if err := v.applyBlock(council, targetNum-1, write); err != nil {
+			return nil, 0, err
 		}
 	} else {
-		var err error
-		council, err = v.getCouncilGenesis()
-		if err != nil {
-			return nil, 0, err
+		if v.validatorVoteBlockNumsCache == nil {
+			v.validatorVoteBlockNumsCache = ReadValidatorVoteBlockNums(v.ChainKv)
+		}
+		nums := v.validatorVoteBlockNumsCache
+		if nums == nil {
+			return nil, 0, errNoVoteBlockNums
+		}
+		pMinVoteNum := ReadLowestScannedVoteNum(v.ChainKv)
```

Comment:
Make `ReadLowestScannedVoteNumCached` so we can share it with `getCouncilDB`

Reply:
You mean the read function for lowestScannedVoteNum, right? made a cache for lowestScannedVoteNum and added the read function for it. https://github.com/kaiachain/kaia/pull/349/commits/9763653a35a197c32d33f842b2b662f26ec92f79

Reply:
made a read function for it! https://github.com/kaiachain/kaia/pull/349/commits/7ab953b3a20f1d712a368be49eda5cb594858d83